tools/assets-automation/assets-maintenance-tool/Azure.Sdk.Tools.Assets.MaintenanceTool/Scan/AssetsScanner.cs

```diff
@@ -0,0 +1,359 @@
+using System;
+using System.IO;
+using System.Text;
+using System.Text.Json;
+using Azure.Sdk.Tools.Assets.MaintenanceTool.Model;
+using Azure.Sdk.Tools.TestProxy.Common;
+using Azure.Sdk.Tools.TestProxy.Common.Exceptions;
+using Azure.Sdk.Tools.TestProxy.Store;
+using Microsoft.Extensions.FileSystemGlobbing;
+
+namespace Azure.Sdk.Tools.Assets.MaintenanceTool.Scan
+{
+    /// <summary>
+    /// Used to walk through repo configurations and locate all assets.
+    /// </summary>
+    public class AssetsScanner
+    {
+        public string WorkingDirectory { get; set; }
+        public static readonly string GIT_TOKEN_ENV_VAR = "GIT_TOKEN";
+
+        private string ResultsFile {
+            get {
+                return Path.Combine(WorkingDirectory, "output.json");
+            }
+        }
+
+        public GitProcessHandler handler { get; set; } = new GitProcessHandler();
+
+        public AssetsScanner(string workingDirectory)
+        {
+            WorkingDirectory = workingDirectory;
+        }
+
+        public AssetsScanner()
+        {
+            WorkingDirectory = Directory.GetCurrentDirectory();
+        }
+
+        /// <summary>
+        /// Walk a run configuration an create a resultSet of all found assets.json references.
+        /// 
+        /// This function automatically takes previous output into account by checking in the current
+        /// working directory for an "output.json" file.
+        /// </summary>
+        /// <param name="config"></param>
+        /// <returns></returns>
+        public AssetsResultSet Scan(RunConfiguration config)
+        {
+            var resultSet = new List<AssetsResult>();
+            var existingResults = ParseExistingResults();
+
+            Parallel.ForEach(config.Repos, repoConfig =>
+            {
+                resultSet.AddRange(ScanRepo(repoConfig, existingResults));
+            });
+
+            var newResults = new AssetsResultSet(resultSet);
+
+            Save(newResults);
+
+            return newResults;
+        }
+
+        /// <summary>
+        /// If the tool is invoked in a directory containing an `"output.json" file, that file will be parsed
+        /// for it's results. The file itself is merely a List of type AssetsResult serialized to disk.
+        /// </summary>
+        /// <returns></returns>
+        public AssetsResultSet? ParseExistingResults()
+        {
+            if (File.Exists(ResultsFile))
```

Comment:
In the spirit of functional programming and explicit I/O, I would obtain ResultsFile only once at the beginning of execution chain, and then always pass it explicitly to all called functions, including ParseExistingResults, etc. But that's just me ;)